third_party/move/move-vm/runtime/src/frame.rs

```diff
@@ -140,24 +152,54 @@ impl Frame {
         frame_cache: Rc<RefCell<FrameTypeCache>>,
     ) -> PartialVMResult<Frame> {
         let ty_args = function.ty_args();
-        for ty in function.local_tys() {
-            gas_meter
-                .charge_create_ty(NumTypeNodes::new(ty.num_nodes_in_subst(ty_args)? as u64))?;
-        }
 
         let ty_builder = vm_config.ty_builder.clone();
-        let local_tys = if RTTCheck::should_perform_checks(&function.function) {
-            if ty_args.is_empty() {
-                function.local_tys().to_vec()
+        let local_tys = if ty_args.is_empty() {
+            // Function is not generic - avoid cloning types.
+            for ty in function.local_tys() {
+                gas_meter.charge_create_ty(NumTypeNodes::new(ty.num_nodes() as u64))?;
+            }
+
+            if RTTCheck::should_perform_checks(&function.function) {
+                LocalTys::BorrowFromFunction
             } else {
-                function
-                    .local_tys()
-                    .iter()
-                    .map(|ty| ty_builder.create_ty_with_subst(ty, ty_args))
-                    .collect::<PartialVMResult<Vec<_>>>()?
+                LocalTys::None
             }
         } else {
-            vec![]
+            // Try cached instantiated locals in frame cache. This way we instantiate only once per
+            // usage of the function.
+            let mut cache_borrow = frame_cache.borrow_mut();
```

Comment:
I wonder whether what you are doing here belongs into the frame cache, passing the gas meter in?